Hlb update and document external build steps (#2769)
digital-asset/daml: language-support/hs/bindings/README.md

As well as `stack`, the only other prerequisite is to have `grpc` installed. (`grpc` is not required to generate the `daml-ledger` package, but it is required to use it).

We are currently using gRPC version `1.23.0`. To install `grpc`  requires building `grpc` from source (really!). See detailed instructions here: https://github.com/grpc/grpc/blob/master/BUILDING.md.

In the instructions below we do `make install` for grpc despite the warning from the `grpc` build instructions:

> *"WARNING: After installing with make install there is no easy way to uninstall, which can cause issues if you later want to remove the grpc and/or protobuf installation or upgrade to a newer version."*

If you decide against the `make install`, or choose a different install location, you will need to adjust the settings of `extra-lib-dirs` and `extra-include-dirs` in your `stack.yaml` config.

Also, in the instructions below we export the `daml-ledger` package to `/tmp` which matches the locat

> ion declared in the `stack.yaml` of the example application `nim`. If you export somewhere else, you will need to adapt your `stack.yaml`

Reviewer: Rendom extra newline here in the middle of `location`.

Author: nice spot, thanks

## Download and build grpc at version 1.23.0

    git clone -b v1.23.0 https://github.com/grpc/grpc
    cd grpc
    git submodule update --init
    make
    make prefix=/usr/local/grpc install

## Clone daml repo, and export the daml-ledger package

    cd /tmp
    git clone https://github.com/digital-asset/daml.git
    cd daml

>     direnv allow

Reviewer: ```suggestion
```
The next line takes care of initializing the environment, there is no need to do both.

Author: I wasn't sure. Removed it now.

[here](/language-support/hs/bindings/test/DA/Ledger/Tests.hs) and
[here](/language-support/hs/bindings/examples/chat/src/DA/Ledger/App/Chat/ChatLedger.hs).

# Using these bindings **outside of this repository**

*Note: These instructions document the current process. We do plan make it simpler!*

These bindings can be exported as a standalone Haskell package `daml-ledger` using `stack`. Currently, `stack` references the code directly in the `daml` repo. The only annoyance is that part of the Haskell code for `daml-ledger` is generated from `.proto` files, and this generation step must be performed using the `bazel` build. We plan to simplify this step by directly providing `daml-ledger` as a package on hackage. Instructions for working with the daml repo can be found here: https://github.com/digital-asset/daml

> As well as `stack`, the only other prerequisite is to have `grpc` installed.

Author: ok, did this

We are currently using gRPC version `1.23.0`. To install `grpc`  requires building `grpc` from source (really!). See detailed instructions here: https://github.com/grpc/grpc/blob/master/BUILDING.md.

Note, the `grpc` build instructions warn against doing a global install:

> *"WARNING: After installing with make install there is no easy way to uninstall, which can cause issues if you later want to remove the grpc and/or protobuf installation or upgrade to a newer version."*

If you follow this advice, then you must adjust `extra-lib-dirs` in your `stack.yaml` config. This is explicated in the detailed instructions below.

## Download (somewhere permanent) and build grpc at version 1.23.0

    git clone -b v1.23.0 https://github.com/grpc/grpc
    cd grpc
    GRPC=$(pwd)                         # for later
    git submodule update --init

>     make

Author: Yeah. I changed the process and instructions to do a `make install`
It makes things much simpler!
I chose `/usr/local/grpc` to match the `stack.yaml`
And this way there is no need for the `sed` step in the instructions.

>     make

Reviewer: Touching `/usr/local` does require root access on many systems though :(

>     make

Author: I know. But they don't have to follow the instructions exactly.

## Clone daml repo; and export the daml-ledger package

    cd /tmp
    git clone https://github.com/digital-asset/daml.git
    cd daml
    direnv allow
    language-support/hs/bindings/export-package.sh /tmp

## Write a DAML Ledger App in Haskell (or copy one!)

    cd /tmp
    cp -rp /tmp/daml/language-support/hs/bindings/examples/nim nim
    cd nim

## Adjust `stack.yaml` to find `grpc`

>     sed -i 's,/usr/local/grpc/lib,'$GRPC'/libs/opt,' stack.yaml

Author: Yes, double quotes are simpler.  But I removed the `sed` step now

## Build

>     stack build

Author: Yes. did this

[here](/language-support/hs/bindings/test/DA/Ledger/Tests.hs) and
[here](/language-support/hs/bindings/examples/chat/src/DA/Ledger/App/Chat/ChatLedger.hs).

# Using these bindings **outside of this repository**

*Note: These instructions document the current process. We do plan make it simpler!*

These bindings can be build as a standalone Haskell package `daml-ledger` using `stack`. Currently, `stack` references the code directly in the `daml` repo. The only annoyance is that part of the Haskell code for `daml-ledger` is generated from `.proto` files, and this generation step must be performed using the `bazel` build. We plan to simplify this step by directly providing `daml-ledger` as a package on hackage.

As well as `stack`, the only other prerequisite is to have `grpc` installed.

> We are currently using gRPC version `1.22.1`. To install `grpc`  requires building `grpc` from source (really!). See detailed instructions here: https://github.com/grpc/grpc/blob/master/BUILDING.md.

Reviewer 2: ```suggestion
We are currently using gRPC version `1.23.0`. To install `grpc`  requires building `grpc` from source (really!). See detailed instructions here: https://github.com/grpc/grpc/blob/master/BUILDING.md.
```

Author: done

Note, the `grpc` build instructions warn against doing a global install:

> *"WARNING: After installing with make install there is no easy way to uninstall, which can cause issues if you later want to remove the grpc and/or protobuf installation or upgrade to a newer version."*

If you follow this advice, then you must adjust `extra-lib-dirs` in your `stack.yaml` config. This is explicated in the detailed instructions below.

> ## Download and build grpc at version 1.22.1

Reviewer 2: ```suggestion
## Download and build grpc at version 1.23.0
```

>     cd ~/github

Reviewer 2: That line doesn’t seem useful. I am also not sure we really want to maintain our own instructions here. What’s wrong with the ones in the gRPC repo?

Author: I got rid of `cd ~/github`
I just made the instructions say to build `grpc` somewhere permanent.

Author: I think it's useful to include the `grpc` build step in our instructions.

>     git clone -b $(curl -L https://grpc.io/release) https://github.com/grpc/grpc

Reviewer: ```suggestion
    git clone -b v1.22.1 https://github.com/grpc/grpc
```
and then you don't need the additional `git checkout` step below

Author: done

    cd grpc

>     git checkout v1.22.1

Reviewer 2: ```suggestion
    git checkout v1.23.0
```

Author: done

>     git checkout v1.22.1

Reviewer: ```suggestion
```

    git submodule update --init
    make

## Clone daml repo; force Haskell code generation

>     cd ~/github

Reviewer 2: Same here, this line doesn’t add anything.

Author: I changed the instructions to clone `daml` in `/tmp'.
I think this simplicity is nice.

    git clone https://github.com/digital-asset/daml.git
    cd daml

>     direnv allow

Reviewer 2: I would just link to the dev-env setup instructions in our main readme.

Author: ok, I did this

>     direnv allow

Reviewer: most people will not have `direnv` installed, so probably better to just enale the dev-env once as we do on CI
```suggestion
    bash
    eval $(dev-env/bin/dade-assist)
```
(extra `bash` to create a new shell, so they can then undo the `dade assist` call by typing exit; it may be worth explaining that in text around the code block)

Author: I tried this, but I still needed `direnv allow`.

Author: so I didn't make the change you suggested

>     bazel build ledger-api/grpc-definitions:google-protobuf-haskellpb-sources ledger-api/grpc-definitions:google-rpc-haskellpb-sources ledger-api/grpc-definitions:ledger-api-haskellpb-sources ledger-api/grpc-definitions:ledger-api-haskellpb-sources-admin ledger-api/grpc-definitions:ledger-api-haskellpb-sources-testing

Reviewer 2: Moving from a bash script to having to run this yourself seems worse. Why can’t we automate this?

Author: I reinstated a script to do the required `bazel` build and export the (now single) dist tarball.

>     bazel build ledger-api/grpc-definitions:google-protobuf-haskellpb-sources ledger-api/grpc-definitions:google-rpc-haskellpb-sources ledger-api/grpc-definitions:ledger-api-haskellpb-sources ledger-api/grpc-definitions:ledger-api-haskellpb-sources-admin ledger-api/grpc-definitions:ledger-api-haskellpb-sources-testing

Reviewer: ```suggestion
    bazel build ledger-api/grpc-definitions:google-protobuf-haskellpb-sources ledger-api/grpc-definitions:google-rpc-haskellpb-sources ledger-api/grpc-definitions:ledger-api-haskellpb-sources ledger-api/grpc-definitions:ledger-api-haskellpb-sources-admin ledger-api/grpc-definitions:ledger-api-haskellpb-sources-testing
    exit
```

Author: didn't do the `bash` ... `exit`

## Write a DAML Ledger App in Haskell (or copy one!)

>     cd /tmp

Reviewer: ```suggestion
    cd $(mktemp -d)
```

Author: I played around with using `$(mktemp -d)` in the instuctions... but in the end it seemed to make things less clear, so I went back to `/tmp'

But overall, the instructions are simpler than before.

The instructions no longer have a step to patch the `stack.yaml` copied from the `nim` example with the location of the `daml-ledger` package... because the original file assumes the package has been exported to `/tmp`, and that matches where the instructions export it to!
